src/environmentd/src/http/static/js/memory.jsx

```diff
@@ -27,52 +27,148 @@ async function query(sql) {
+    window.history.replaceState({}, '', `${location.pathname}?${params}`);
+  }, [currentClusterName, currentReplicaName])
+
+  return (
+    <div>
+      {loading ? (
+        <div>Loading...</div>
+      ) : error ? (
+        <div>error: {error}</div>
+      ) : (
+        <div>
+          <label htmlFor="cluster_replica">Cluster Replica </label>
+          <select
+            id="cluster_replica"
+            name="cluster_replica"
+            onChange={(event) => {
+              const clusterReplica = event.target.value;
+              const clusterReplicaSplit = clusterReplica.split('|');
+              setCurrentClusterName(clusterReplicaSplit[0]);
+              setCurrentReplicaName(clusterReplicaSplit[1]);
+            }}
+            defaultValue={`${currentClusterName}|${currentReplicaName}`}>
+            {sqlResponse.map((v) => (
+              <option key={`${v[0]}|${v[1]}`} value={`${v[0]}|${v[1]}`}>
+                {`${v[0]}.${v[1]}`}
+              </option>
+            ))}
+          </select>
+          <Views clusterName={currentClusterName} replicaName={currentReplicaName} />
+        </div>
+      )}
+    </div>
+  );
 }
 
-function Views() {
+function Views(props) {
+  const [currentDataflow, setCurrentDataflow] = useState(null);
+  const [includeSystemCatalog, setIncludeSystemCatalog] = useState(false);
+  const [records, setRecords] = useState(null);
+  const [loading, setLoading] = useState(true);
+  const [error, setError] = useState(false);
+
   useEffect(() => {
     const search = new URLSearchParams(location.search);
     const dataflow = search.get('dataflow');
     if (dataflow) {
-      setCurrent(dataflow);
+      setCurrentDataflow(dataflow);
     }
     setIncludeSystemCatalog(search.get('system_catalog') === 'true');
   }, []);
 
-  const [current, setCurrent] = useState(null);
-  const [includeSystemCatalog, setIncludeSystemCatalog] = useState(false);
+  useEffect(() => {
+    setCurrentDataflow(null);
+  }, [props]);
 
-  const where_fragment = includeSystemCatalog ? `` : `WHERE name NOT LIKE 'Dataflow: mz_catalog.%'`;
+  useEffect(() => {
+    const params = new URLSearchParams(location.search);
+    if (currentDataflow) {
+      params.set('dataflow', currentDataflow);
+    } else {
+      params.delete('dataflow');
+    }
+    window.history.replaceState({}, '', `${location.pathname}?${params}`);
+  }, [currentDataflow]);
 
-  const queryMaterializedViews = `
-    SELECT
-      id, name, records
-    FROM
-      mz_internal.mz_records_per_dataflow_global
-    ${where_fragment}
-    ORDER BY
-      records DESC
-  `;
+  const whereFragment = includeSystemCatalog ? `` : `WHERE name NOT LIKE 'Dataflow: mz_catalog.%'`;
+
+  useEffect(() => {
+    setLoading(true);
+    setError(false);
+
+    const load = async () => {
+      const {
+        results: [_set_cluster, _set_replica, records_table],
+      } = await query(`
+        SET cluster = ${props.clusterName};
```

Comment:
this will break if a name contains a space, single or double quote (maybe just one of those, unsure). SET statements don't appear to accept params, so you'll need to escape them. I think you can do this always wrapping the names in single quotes, and replacing any single quote in the name with two single quotes.

Reply:
Seems to be any valid identifier: `create cluster "'"" " replicas( s (size '1'));`

Reply:
Yeah I tested locally with a bunch of odd names, e.g. `test sp`, `test'qu` and `test"qu`, and the formatting I added handled them.